libs/api/ebpf_api.cpp

```diff
@@ -19,6 +21,17 @@ extern "C"
 
 #define MAX_CODE_SIZE (32 * 1024) // 32 KB
 
+static uint64_t _ebpf_file_descriptor_counter = 0;
+static std::map<fd_t, ebpf_program_t*> _ebpf_programs;
+static std::map<fd_t, ebpf_map_t*> _ebpf_maps;
+static std::vector<ebpf_object_t*> _ebpf_objects;
+
+#define for_each_program_in_object(x) for (entry = x->programs.Flink; entry != &x->programs; entry = entry->Flink)
+#define for_each_map_in_object(x) for (entry = x->maps.Flink; entry != &x->maps; entry = entry->Flink)
```

Comment:
why use ebpf_list_entry_t walks instead of making x->programs and x->maps be vectors that you can do simply C++ iteration over?

Reply:
Some APIs (like ebpf_program_next(), ebpf_program_previous(), ect.) have to return pointers to ebpf_map and ebpf_program structures. We should ideally not return pointers to vector elements, as those pointers can become invalid.

Reply:
iterable is not part of the element itself it's a wrapper. And you can use unique_ptr or shared_ptr as appropriate to manage pointer lifetime.

Reply:
A problem with using vectors is that the APIs like ebpf_program_next()/ebpf_program_previous() will not remain O(1) (unless we also save the vector index in the ebpf_program / ebpf_map structures).

Reply:
What does Linux/libbpf do?

Vector index is not as good as saving the last key returned, since there may be deletions that change the indices.

O(1) is not really important unless there's a large number of programs.   If we're just talking about 10s of programs then it's negligible.

Reply:
Libbpf uses linked list to save ebpf_program and ebpf_map in ebpf_object structure. But that is just an implementation detail and can be different in this implementation. 

If there is some drawback in using linked list here compared to vectors, I can change the code to use vectors.

Reply:
working on changing to vectors in the next commit.

Reply:
Changed to using vector.